File: 프로그래머스/3/42892. 길 찾기 게임/길 찾기 게임.java
import java.util.*;

class Solution {
    static class Node{
        int nodeNum; //노드 번호
        int[] co; //좌표
        Node leftNode;
        Node rightNode;

        Node(int nodeNum, int[] co){
            this.nodeNum=nodeNum;
            this.co=co;
        }
    }

    static class Tree{
        Node root;

        Tree(){}

        void insertNode(int nodeNum, int[] co){
            Node newNode = new Node(nodeNum, co);
            if(this.root==null){
                this.root=newNode;
                return;
            }
            Node current = this.root;
            while(current!=null){
                if(current.co[0]>co[0]){ // 부모노드 기준 왼쪽 노드일 경우
                    if(current.leftNode==null){
                        current.leftNode=newNode;
                        break;
                    }else{
                        current = current.leftNode;
                    }
                }else{ //부모노드 기준 오른쪽 노드일 경우
                    if(current.rightNode==null){
                        current.rightNode = newNode;
                        break;
                    }else{
                        current = current.rightNode;
                    }
                }
            }
        }
    }
    static int preidx, postidx, nodeCnt;
    static int[][] answer;
    
    public void preorder(Node current){ //전위 순회
        if(preidx>=nodeCnt) return;
        answer[0][preidx++] = current.nodeNum;
        if(current.leftNode!=null) preorder(current.leftNode);
        if(current.rightNode!=null) preorder(current.rightNode);
    }
    
    public void postorder(Node current){// 후위 순회
        if(postidx>=nodeCnt) return;
        if(current.leftNode!=null) postorder(current.leftNode);
        if(current.rightNode!=null) postorder(current.rightNode);
        answer[1][postidx++] = current.nodeNum;
    }
    
    public int[][] solution(int[][] nodeinfo) {
        nodeCnt = nodeinfo.length;
        answer = new int[2][nodeCnt];
        Map<String, Integer> nodeMap = new HashMap<>();
        for(int i=0; i<nodeCnt; i++){
           nodeMap.put(nodeinfo[i][0]+","+nodeinfo[i][1], i+1); //key: 좌표, value: 노드 번호
        }
        
        Arrays.sort(nodeinfo, (a, b)->{
           if(a[1]!=b[1]) return b[1]-a[1]; //y축 기준 내림차순 정렬
            return a[0]-b[0]; //x축 기준 오름차순 정렬
        });
        
        Tree tree = new Tree();
        for(int i=0; i<nodeCnt; i++){
            tree.insertNode(nodeMap.get(nodeinfo[i][0]+","+nodeinfo[i][1]), nodeinfo[i].clone());
        }
        
        preidx=0;
        postidx=0;
        
        preorder(tree.root);
        postorder(tree.root);
        
        return answer;
    }
}

File: 프로그래머스/3/42892. 길 찾기 게임/길 찾기 게임.js
class Node {
    constructor(node, co){
        this.node = node; //번호
        this.co = co; //좌표
        this.left = null;
        this.right = null;
    }
}

class Tree{
    constructor(){
        this.root = null;
    }
    
    insertNode(node, co){
        const newNode = new Node(node, co);
        if(this.root === null){ //root 노드 지정
            this.root = newNode;
            return this;
        }
        let current = this.root; //root 노드부터 시작
        while(current){
            if(current.co[0]>co[0]){
                if(current.left === null) { 
                    current.left = newNode;
                    break;
                }else{
                    //이미 왼쪽 노드가 있으면 해당 왼쪽 노드를 부모 노드로 해서 탐색
                    current = current.left; 
                }
            }
            else{
                if(current.right === null) { 
                    current.right = newNode;
                    break;
                }else{
                    current = current.right; 
                }
            }
        }
    }
}

function solution(nodeinfo) {
    let answer = [[],[]];
    const nodeObj = {};
    // 각 노드별 번호 저장. key=좌표, value=노드 번호
    nodeinfo.forEach((e, idx)=>{nodeObj[e[0]+","+e[1]]=idx+1}); 
    // y기준 내림차순, x기준 오름차순 정렬
    nodeinfo.sort((a,b)=>{
        if(a[1]!==b[1]) return b[1]-a[1];
        else return a[0]-b[0];
    });
  
    
    // 이진트리 만들기
    let tree = new Tree();
    for(let i=0; i<nodeinfo.length; i++){
        tree.insertNode(nodeObj[nodeinfo[i][0]+","+nodeinfo[i][1]],nodeinfo[i]);
    }
    
    // 전위 순회 -> answer[0]에 push
    function preorder(current){
        answer[0].push(current.node);
        if(current.left) preorder(current.left);
        if(current.right) preorder(current.right);
    }
    
    // 후위 순회 -> answer[1]에 push
    function postorder(current){
        if(current.left) postorder(current.left);
        if(current.right) postorder(current.right);
        answer[1].push(current.node);
    }
    
    preorder(tree.root);
    postorder(tree.root);
    
    return answer;
}